doc/specs/#190 - Proxy Support.md

```diff
@@ -0,0 +1,93 @@
+---
+author: Flor Chacon @florelis
+created on: 2024-02-07
+last updated: 2024-02-07
+issue id: 190
+---
+
+# Proxy support
+
+For [#190](https://github.com/microsoft/winget-cli/issues/190)
+
+## Abstract
+
+This spec describes a feature to specify a proxy for winget to use when connecting to the internet.
+
+## Solution Design
+
+A new command line argument will be added to specify a proxy to use during a particular invocation of winget.
+This functionality will first need to be enabled through an admin setting, similar to local manifests or hash override.
+
+An option to set a default proxy to use on every flow will be added, but it will require administrator permissions to be set.
+
+New Group Policy will also be added for IT admins to control the use of proxies.
+The policies will be similar to those we already have for sources, so that a specific proxy can be required or only a predefined set of proxies can be allowed.
+
+Proxies will not be used for the configuration features for now.
+
+If a proxy is configured, it will be used for accessing sources and downloading installers.
```

Comment:
I started implementing the feature and just now realized that this is not possible in all cases the way we currently do things. I was assuming it would be basically what @eternalphane did in #1776, i.e., adding the proxy argument in the call to `InternetOpen()`, but we have several other ways of fetching remote content and not all of them expose functionality to specify proxy.

* Delivery Optimization cannot use a custom proxy, only the system-wide one.
* wininet _can_ use a custom proxy (that's `InternetOpen()`), but we basically only use it to download installers.
* The MSIX deployment APIs take a URI and have no option to configure a custom proxy. (I was assuming they took a stream and I could pass the proxied stream, but was wrong..)
* cpprestsdk can take a custom proxy to use
* `Windows.Web.Http.HttpClient` that we use for getting headers (for source update), downloading source MSIX packages and some other things, only allows us to use the system-wide proxy, not to set a custom one.

So, we can implement the feature for REST API requests and for downloading non-msix installers with wininet.
We definitely cannot implement the feature for using DO, or for MSIX installers (unless we download them first, but then we lose streaming installs).
And for the things that use `HttpClient`, we can't unless we decide to re-implement it using wininet or something else.

Would we be comfortable with saying "proxy is only used for non-MSIX installers (where it also forces the use of wininet over DO) and REST sources"? Or how else could we move the feature forward?

Reply:
@florelis these are fantastic call-outs.

I think we should proceed with an experimental feature for proxy with the parts we can implement, and clearly state and link to the remaining details in Issues here (even if they are just to track work external to WinGet so the community can see as they are resolved).

Those other issues should be referenced in "Future Considerations" with links to the issues and we can continue to make progress over time. I have a fairly strong belief that anything we can implement will add value, and it will just continue getting better as we are able to track and close any remaining gaps.